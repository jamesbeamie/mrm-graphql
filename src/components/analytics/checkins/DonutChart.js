import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Doughnut } from 'react-chartjs-2';
import '../../../assets/styles/donutchart.scss';

// Import the tip
import Tip from '../../commons/Tooltip';
import Overlay from '../../commons/Overlay';

// eslint-disable-next-line react/prefer-stateless-function
/**
 * A donut Chart for Checkins
 *
 * @param {array} props
 *
 * @returns {JSX}
 */
class DonutChart extends Component {
  /**
  * Conditionally returns a chart, loader or error message
  *
  * @param {array}
  *
  * @returns {JSX}
  */
  renderChart = () => {
    const {
      loading,
      percentage,
      entries,
      total,
      chartColor,
      hasInfo,
      dataName,
    } = this.props;

    const options = {
      legend: {
        display: false,
      },
      cutoutPercentage: 70,
    };

    let graphData;
    if (total === 0) {
      graphData = {
        labels: [dataName, 'Bookings'],
        datasets: [{
          data: [entries, 1],
          backgroundColor: chartColor,
        }],
        text: '23%',
      };
    } else {
      graphData = {
        labels: [dataName, 'Bookings'],
        datasets: [{
          data: [entries, total],
          backgroundColor: chartColor,
        }],
        text: '23%',
      };
    }

    return (
      <Fragment>
        <div className="diagram">
          <Doughnut
            data={graphData}
            width={160}
            height={160}
            options={options}
          />
          <p className={hasInfo ? 'chart-percentage' : 'no-info-percentage'}>{loading ? 45 : percentage}%</p>
        </div>
        {hasInfo && (
          <div className="chart-text">
            <p>
              {!loading && `${entries}/${total}`}
            </p>
            <p>Meetings</p>
          </div>
        )}
      </Fragment>
    );
  }

  render() {
    const {
      chartTitle,
      hintText,
      loading,
      tip,
    } = this.props;

    return (
      <div className="donut-chart overlay-container">
        {loading && <Overlay />}
        <div className="chart-header">
          <p>{chartTitle}</p>
          <div className="hint">
            {Tip(tip)}
            <div className="hint-text">
              <p>{hintText}</p>
            </div>
          </div>
        </div>
        <div className="chart-content">
          {
            this.renderChart()
          }
        </div>
      </div>
    );
  }
}

DonutChart.propTypes = {
  chartTitle: PropTypes.string,
  hintText: PropTypes.string,
  percentage: PropTypes.number,
  entries: PropTypes.number,
  total: PropTypes.number,
  hasInfo: PropTypes.bool,
  tip: PropTypes.string.isRequired,
  loading: PropTypes.bool,
  chartColor: PropTypes.instanceOf(Object),
  dataName: PropTypes.string,
};
DonutChart.defaultProps = {
  chartTitle: 'Pie Chart',
  hintText: '',
  percentage: 0,
  entries: 0,
  total: 10,
  hasInfo: true,
  loading: null,
  chartColor: null,
  dataName: 'Checkins',
};

export default DonutChart;
