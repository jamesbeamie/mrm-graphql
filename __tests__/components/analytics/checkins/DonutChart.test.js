import React from 'react';
import { shallow } from 'enzyme';
import DonutChart from '../../../../src/components/analytics/checkins/DonutChart';

describe('Checkins component', () => {
  let props;
  let wrapper;

  props = { tip: '', loading: false, isFutureDateSelected: false };
  wrapper = shallow(<DonutChart {...props} />);

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should contain a doughnut chart', () => {
    expect(wrapper.find('.diagram')).toHaveLength(1);
  });

  it('should render properly when loading', () => {
    props = { tip: '', loading: true, isFutureDateSelected: false };
    const wrapperLoading = shallow(<DonutChart {...props} />);
    expect(wrapperLoading.find('Overlay')).toHaveLength(1);
  });

  it('should display the cancellations doughnut chart properly', () => {
    props = { tip: '', hasInfo: false, isFutureDateSelected: false };
    wrapper = shallow(<DonutChart {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
