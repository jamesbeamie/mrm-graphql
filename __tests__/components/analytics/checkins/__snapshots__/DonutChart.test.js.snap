// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkins component should display the cancellations doughnut chart properly 1`] = `
<div
  className="donut-chart overlay-container"
>
  <div
    className="chart-header"
  >
    <p>
      Pie Chart
    </p>
    <div
      className="hint"
    >
      <ThemedTooltippedComponent
        className=""
        composeTheme="deeply"
        mapThemrProps={[Function]}
        theme={null}
        tooltip=""
        tooltipDelay={0}
        tooltipHideOnClick={true}
        tooltipPosition="top"
        tooltipShowOnClick={true}
      >
        <img
          alt="help icon"
          src="https://firebasestorage.googleapis.com/v0/b/learning-map-app.appspot.com/o/MrmFrontImages%2Fhelp_outline_24px.svg?alt=media&token=839e3137-c29e-463d-9a68-361e28794911"
        />
      </ThemedTooltippedComponent>
      <div
        className="hint-text"
      >
        <p />
      </div>
    </div>
  </div>
  <div
    className="chart-content"
  >
    <div
      className="diagram"
    >
      <Doughnut
        data={
          Object {
            "datasets": Array [
              Object {
                "backgroundColor": null,
                "data": Array [
                  0,
                  10,
                ],
              },
            ],
            "labels": Array [
              "Checkins",
              "Bookings",
            ],
            "text": "23%",
          }
        }
        height={160}
        options={
          Object {
            "cutoutPercentage": 70,
            "legend": Object {
              "display": false,
            },
          }
        }
        width={160}
      />
      <p
        className="no-info-percentage"
      >
        0
        %
      </p>
    </div>
  </div>
</div>
`;

exports[`Checkins component should render properly 1`] = `
<div
  className="donut-chart overlay-container"
>
  <div
    className="chart-header"
  >
    <p>
      Pie Chart
    </p>
    <div
      className="hint"
    >
      <ThemedTooltippedComponent
        className=""
        composeTheme="deeply"
        mapThemrProps={[Function]}
        theme={null}
        tooltip=""
        tooltipDelay={0}
        tooltipHideOnClick={true}
        tooltipPosition="top"
        tooltipShowOnClick={true}
      >
        <img
          alt="help icon"
          src="https://firebasestorage.googleapis.com/v0/b/learning-map-app.appspot.com/o/MrmFrontImages%2Fhelp_outline_24px.svg?alt=media&token=839e3137-c29e-463d-9a68-361e28794911"
        />
      </ThemedTooltippedComponent>
      <div
        className="hint-text"
      >
        <p />
      </div>
    </div>
  </div>
  <div
    className="chart-content"
  >
    <div
      className="diagram"
    >
      <Doughnut
        data={
          Object {
            "datasets": Array [
              Object {
                "backgroundColor": null,
                "data": Array [
                  0,
                  10,
                ],
              },
            ],
            "labels": Array [
              "Checkins",
              "Bookings",
            ],
            "text": "23%",
          }
        }
        height={160}
        options={
          Object {
            "cutoutPercentage": 70,
            "legend": Object {
              "display": false,
            },
          }
        }
        width={160}
      />
      <p
        className="chart-percentage"
      >
        0
        %
      </p>
    </div>
    <div
      className="chart-text"
    >
      <p>
        0/10
      </p>
      <p>
        Meetings
      </p>
    </div>
  </div>
</div>
`;
